tests/test_proxy_hwm.cpp

```diff
@@ -0,0 +1,426 @@
+} proxy_hwm_cfg_t;
+
+static void lower_tcp_buff (void *sock_)
+{
+    int sndBuff;
+    size_t sndBuffSz = sizeof sndBuff;
+    int rc = zmq_getsockopt (sock_, ZMQ_SNDBUF, &sndBuff, &sndBuffSz);
+    assert (rc == 0);
+
+    int newBuff = 1000;
+    TEST_ASSERT_SUCCESS_ERRNO (
+      zmq_setsockopt (sock_, ZMQ_SNDBUF, &newBuff, sizeof (newBuff)));
+
+    rc = zmq_getsockopt (sock_, ZMQ_SNDBUF, &sndBuff, &sndBuffSz);
+    assert (rc == 0);
+}
+
+static void lower_hwm (void *skt)
+{
+    int send_hwm_ = HWM;
+    TEST_ASSERT_SUCCESS_ERRNO (
+      zmq_setsockopt (skt, ZMQ_SNDHWM, &send_hwm_, sizeof (send_hwm_)));
+
+    TEST_ASSERT_SUCCESS_ERRNO (
+      zmq_setsockopt (skt, ZMQ_RCVHWM, &send_hwm_, sizeof (send_hwm_)));
+}
+
+static void publisher_thread_main (void *pvoid)
+{
+    proxy_hwm_cfg_t *cfg = (proxy_hwm_cfg_t *) pvoid;
+
+    void *pubsocket = zmq_socket (cfg->context, ZMQ_PUB);
+    assert (pubsocket);
+
+    lower_hwm (pubsocket);
+
+    TEST_ASSERT_SUCCESS_ERRNO (zmq_connect (pubsocket, cfg->frontend_endpoint));
+
+    int optval = 1;
+    TEST_ASSERT_SUCCESS_ERRNO (
+      zmq_setsockopt (pubsocket, ZMQ_XPUB_NODROP, &optval, sizeof (optval)));
+
+    msleep (SETTLE_TIME);
+
+    uint64_t send_count = 0;
+    while (true) {
+        zmq_msg_t msg;
+        int rc = zmq_msg_init_size (&msg, NUM_BYTES_PER_MSG);
+        assert (rc == 0);
+
+        /* Fill in message content with 'AAAAAA' */
+        memset (zmq_msg_data (&msg), 'A', NUM_BYTES_PER_MSG);
+
+        /* Send the message to the socket */
+        rc = zmq_msg_send (&msg, pubsocket, ZMQ_DONTWAIT);
+        if (rc != -1) {
+            send_count++;
+        } else {
+            TEST_ASSERT_SUCCESS_ERRNO (zmq_msg_close (&msg));
+            break;
+        }
+    }
+
+    // VERIFY EXPECTED RESULTS
+
+    TEST_ASSERT (4 * HWM == send_count || 2 * HWM == send_count);
+
+
+    // CLEANUP
+
+    zmq_close (pubsocket);
+}
+
+static void subscriber_thread_main (void *pvoid)
+{
+    proxy_hwm_cfg_t *cfg = (proxy_hwm_cfg_t *) pvoid;
+
+    void *subsocket = zmq_socket (cfg->context, ZMQ_SUB);
+    assert (subsocket);
+
+    lower_hwm (subsocket);
+
+    TEST_ASSERT_SUCCESS_ERRNO (zmq_setsockopt (subsocket, ZMQ_SUBSCRIBE, 0, 0));
+
+    TEST_ASSERT_SUCCESS_ERRNO (zmq_connect (subsocket, cfg->backend_endpoint));
+
+    lower_tcp_buff (subsocket);
+
+    // receive all sent messages
+    uint64_t rxsuccess = 0;
+    bool success = true;
+    while (success) {
+        zmq_msg_t msg;
+        int rc = zmq_msg_init (&msg);
+        assert (rc == 0);
+
+        rc = zmq_msg_recv (&msg, subsocket, 0);
+        if (rc != -1) {
+            rxsuccess++;
+
+            // after receiving 1st message, set a finite timeout (default is infinite)
+            int timeout_ms = 100;
+            TEST_ASSERT_SUCCESS_ERRNO (zmq_setsockopt (
+              subsocket, ZMQ_RCVTIMEO, &timeout_ms, sizeof (timeout_ms)));
+        } else {
+            TEST_ASSERT_SUCCESS_ERRNO (zmq_msg_close (&msg));
```

Comment:
this close call needs to be in the other branch of the if, that's what's causing the leak

Reply:
you're right - my fault.. I intended to puth the close on the success path indeed